File: packages/alpaca-time/src/calendar.ts
import { fail } from './error';
import { now, parts, today } from './clock';
import {
  addUtcDays,
  datePartsToUtcDate,
  formatDateParts,
  parseDateParts,
  utcDate,
  utcDateToDateParts,
} from './internal';
import type { MarketHours, NyDateString, NyTimestampString, TradingDayInfo } from './types';

const MONDAY = 1;
const THURSDAY = 4;
const SATURDAY = 6;
const SUNDAY = 0;

const holidayCache = new Map<number, Set<string>>();
const earlyCloseCache = new Map<number, Set<string>>();

function observedFixedHoliday(year: number, month: number, day: number): Date {
  const date = utcDate(year, month, day);
  const weekday = date.getUTCDay();
  if (weekday === SATURDAY) {
    return addUtcDays(date, -1);
  }
  if (weekday === SUNDAY) {
    return addUtcDays(date, 1);
  }
  return date;
}

function nthWeekday(year: number, month: number, weekday: number, nth: number): Date {
  let current = utcDate(year, month, 1);
  while (current.getUTCDay() !== weekday) {
    current = addUtcDays(current, 1);
  }
  return addUtcDays(current, (nth - 1) * 7);
}

function lastWeekday(year: number, month: number, weekday: number): Date {
  const nextMonth = month === 12 ? utcDate(year + 1, 1, 1) : utcDate(year, month + 1, 1);
  let current = addUtcDays(nextMonth, -1);
  while (current.getUTCDay() !== weekday) {
    current = addUtcDays(current, -1);
  }
  return current;
}

function easterDate(year: number): Date {
  const a = year % 19;
  const b = Math.floor(year / 100);
  const c = year % 100;
  const d = Math.floor(b / 4);
  const e = b % 4;
  const f = Math.floor((b + 8) / 25);
  const g = Math.floor((b - f + 1) / 3);
  const h = (19 * a + b - d - g + 15) % 30;
  const i = Math.floor(c / 4);
  const k = c % 4;
  const l = (32 + 2 * e + 2 * i - h - k) % 7;
  const m = Math.floor((a + 11 * h + 22 * l) / 451);
  const month = Math.floor((h + l - 7 * m + 114) / 31);
  const day = ((h + l - 7 * m + 114) % 31) + 1;
  return utcDate(year, month, day);
}

function isWeekday(date: Date): boolean {
  const weekday = date.getUTCDay();
  return weekday !== SATURDAY && weekday !== SUNDAY;
}

function holidayDatesForYear(year: number): Set<string> {
  const cached = holidayCache.get(year);
  if (cached) {
    return cached;
  }

  const dates = new Set<string>([
    formatDateParts(utcDateToDateParts(observedFixedHoliday(year, 1, 1))),
    formatDateParts(utcDateToDateParts(nthWeekday(year, 1, MONDAY, 3))),
    formatDateParts(utcDateToDateParts(nthWeekday(year, 2, MONDAY, 3))),
    formatDateParts(utcDateToDateParts(addUtcDays(easterDate(year), -2))),
    formatDateParts(utcDateToDateParts(lastWeekday(year, 5, MONDAY))),
    formatDateParts(utcDateToDateParts(observedFixedHoliday(year, 6, 19))),
    formatDateParts(utcDateToDateParts(observedFixedHoliday(year, 7, 4))),
    formatDateParts(utcDateToDateParts(nthWeekday(year, 9, MONDAY, 1))),
    formatDateParts(utcDateToDateParts(nthWeekday(year, 11, THURSDAY, 4))),
    formatDateParts(utcDateToDateParts(observedFixedHoliday(year, 12, 25))),
  ]);

  holidayCache.set(year, dates);
  return dates;
}

function isMarketHolidayDate(date: Date): boolean {
  return holidayDatesForYear(date.getUTCFullYear()).has(formatDateParts(utcDateToDateParts(date)));
}

function isTradingDateDate(date: Date): boolean {
  return isWeekday(date) && !isMarketHolidayDate(date);
}

function addTradingDaysFrom(date: Date, days: number): Date {
  let current = new Date(date.getTime());
  const step = days === 0 ? 0 : days > 0 ? 1 : -1;
  let remaining = Math.abs(days);
  if (step === 0) {
    return current;
  }
  while (remaining > 0) {
    current = addUtcDays(current, step);
    if (isTradingDateDate(current)) {
      remaining -= 1;
    }
  }
  return current;
}

function earlyCloseDatesForYear(year: number): Set<string> {
  const cached = earlyCloseCache.get(year);
  if (cached) {
    return cached;
  }

  const dates = new Set<string>();
  const thanksgiving = nthWeekday(year, 11, THURSDAY, 4);
  const blackFriday = addUtcDays(thanksgiving, 1);
  if (isWeekday(blackFriday) && !isMarketHolidayDate(blackFriday)) {
    dates.add(formatDateParts(utcDateToDateParts(blackFriday)));
  }

  const independenceHoliday = observedFixedHoliday(year, 7, 4);
  const independenceEve = addTradingDaysFrom(independenceHoliday, -1);
  if (independenceEve.getUTCFullYear() === year) {
    dates.add(formatDateParts(utcDateToDateParts(independenceEve)));
  }

  const christmasEve = utcDate(year, 12, 24);
  if (isWeekday(christmasEve) && !isMarketHolidayDate(christmasEve)) {
    dates.add(formatDateParts(utcDateToDateParts(christmasEve)));
  }

  earlyCloseCache.set(year, dates);
  return dates;
}

function isEarlyCloseDate(date: Date): boolean {
  return earlyCloseDatesForYear(date.getUTCFullYear()).has(formatDateParts(utcDateToDateParts(date)));
}

function validateDays(days: number): void {
  if (!Number.isInteger(days)) {
    fail('invalid_days', `days must be integer: ${days}`);
  }
}

export function tradingDayInfo(date: string): TradingDayInfo {
  const parsed = datePartsToUtcDate(parseDateParts(date));
  const canonicalDate = formatDateParts(utcDateToDateParts(parsed));
  const is_trading_date = isTradingDateDate(parsed);
  const is_market_holiday = isMarketHolidayDate(parsed);
  const is_early_close = is_trading_date && isEarlyCloseDate(parsed);
  const market_hours = marketHoursForDate(canonicalDate);

  return {
    date: canonicalDate,
    is_trading_date,
    is_market_holiday,
    is_early_close,
    market_hours,
  };
}

export function isTradingDate(date: string): boolean {
  try {
    return isTradingDateDate(datePartsToUtcDate(parseDateParts(date)));
  } catch {
    return false;
  }
}

export function isTradingToday(): boolean {
  return isTradingDate(today());
}

export function marketHoursForDate(date: string): MarketHours {
  const parsed = datePartsToUtcDate(parseDateParts(date));
  const canonicalDate = formatDateParts(utcDateToDateParts(parsed));
  const is_trading_date = isTradingDateDate(parsed);
  const is_early_close = is_trading_date && isEarlyCloseDate(parsed);

  if (!is_trading_date) {
    return {
      date: canonicalDate,
      is_trading_date: false,
      is_early_close: false,
      premarket_open: null,
      regular_open: null,
      regular_close: null,
      after_hours_close: null,
    };
  }

  return {
    date: canonicalDate,
    is_trading_date: true,
    is_early_close,
    premarket_open: '04:00',
    regular_open: '09:30',
    regular_close: is_early_close ? '13:00' : '16:00',
    after_hours_close: '20:00',
  };
}

export function lastCompletedTradingDate(atTimestamp?: NyTimestampString): NyDateString {
  const parsed = parts(atTimestamp ?? now());
  const hours = marketHoursForDate(parsed.date);

  if (!hours.is_trading_date) {
    return addTradingDays(parsed.date, -1);
  }

  if (!hours.regular_close) {
    fail('invalid_market_hours', 'missing regular close');
  }

  const currentMinutes = parsed.hour * 60 + parsed.minute;
  const closeMinutes = Number(hours.regular_close.slice(0, 2)) * 60 + Number(hours.regular_close.slice(3, 5));

  return currentMinutes >= closeMinutes ? parsed.date : addTradingDays(parsed.date, -1);
}

export function addTradingDays(date: string, days: number): NyDateString {
  validateDays(days);
  const parsed = datePartsToUtcDate(parseDateParts(date));
  return formatDateParts(utcDateToDateParts(addTradingDaysFrom(parsed, days)));
}

File: crates/alpaca-time/src/calendar.rs
use chrono::{Datelike, Duration, NaiveDate, Timelike, Weekday};
use std::collections::HashSet;

use crate::clock::{minutes_from_hhmm, now, parse_naive_date, parse_naive_timestamp, today};
use crate::error::{TimeError, TimeResult};
use crate::types::{MarketHours, TradingDayInfo};

fn observed_fixed_holiday(year: i32, month: u32, day: u32) -> NaiveDate {
    let date = NaiveDate::from_ymd_opt(year, month, day).expect("valid fixed holiday");
    match date.weekday() {
        Weekday::Sat => date - Duration::days(1),
        Weekday::Sun => date + Duration::days(1),
        _ => date,
    }
}

fn nth_weekday(year: i32, month: u32, weekday: Weekday, nth: u32) -> NaiveDate {
    let mut current = NaiveDate::from_ymd_opt(year, month, 1).expect("valid month");
    while current.weekday() != weekday {
        current += Duration::days(1);
    }
    current + Duration::days(((nth - 1) * 7) as i64)
}

fn last_weekday(year: i32, month: u32, weekday: Weekday) -> NaiveDate {
    let next_month = if month == 12 {
        NaiveDate::from_ymd_opt(year + 1, 1, 1).unwrap()
    } else {
        NaiveDate::from_ymd_opt(year, month + 1, 1).unwrap()
    };
    let mut current = next_month - Duration::days(1);
    while current.weekday() != weekday {
        current -= Duration::days(1);
    }
    current
}

fn easter_date(year: i32) -> NaiveDate {
    let a = year % 19;
    let b = year / 100;
    let c = year % 100;
    let d = b / 4;
    let e = b % 4;
    let f = (b + 8) / 25;
    let g = (b - f + 1) / 3;
    let h = (19 * a + b - d - g + 15) % 30;
    let i = c / 4;
    let k = c % 4;
    let l = (32 + 2 * e + 2 * i - h - k) % 7;
    let m = (a + 11 * h + 22 * l) / 451;
    let month = (h + l - 7 * m + 114) / 31;
    let day = ((h + l - 7 * m + 114) % 31) + 1;
    NaiveDate::from_ymd_opt(year, month as u32, day as u32).unwrap()
}

fn holiday_dates_for_year(year: i32) -> HashSet<NaiveDate> {
    let mut holidays = HashSet::new();
    holidays.insert(observed_fixed_holiday(year, 1, 1));
    holidays.insert(nth_weekday(year, 1, Weekday::Mon, 3));
    holidays.insert(nth_weekday(year, 2, Weekday::Mon, 3));
    holidays.insert(easter_date(year) - Duration::days(2));
    holidays.insert(last_weekday(year, 5, Weekday::Mon));
    holidays.insert(observed_fixed_holiday(year, 6, 19));
    holidays.insert(observed_fixed_holiday(year, 7, 4));
    holidays.insert(nth_weekday(year, 9, Weekday::Mon, 1));
    holidays.insert(nth_weekday(year, 11, Weekday::Thu, 4));
    holidays.insert(observed_fixed_holiday(year, 12, 25));
    holidays
}

fn early_close_dates_for_year(year: i32) -> HashSet<NaiveDate> {
    let mut early_close = HashSet::new();

    let thanksgiving = nth_weekday(year, 11, Weekday::Thu, 4);
    let black_friday = thanksgiving + Duration::days(1);
    if is_weekday(black_friday) && !holiday_dates_for_year(year).contains(&black_friday) {
        early_close.insert(black_friday);
    }

    let independence_holiday = observed_fixed_holiday(year, 7, 4);
    let independence_eve = add_trading_days_from(independence_holiday, -1);
    if independence_eve.year() == year {
        early_close.insert(independence_eve);
    }

    if let Some(christmas_eve) = NaiveDate::from_ymd_opt(year, 12, 24) {
        if is_weekday(christmas_eve) && !holiday_dates_for_year(year).contains(&christmas_eve) {
            early_close.insert(christmas_eve);
        }
    }

    early_close
}

fn is_weekday(date: NaiveDate) -> bool {
    !matches!(date.weekday(), Weekday::Sat | Weekday::Sun)
}

fn add_trading_days_from(mut date: NaiveDate, days: i32) -> NaiveDate {
    let step = match days.cmp(&0) {
        std::cmp::Ordering::Less => -1,
        std::cmp::Ordering::Equal => return date,
        std::cmp::Ordering::Greater => 1,
    };
    let mut remaining = days.abs();
    while remaining > 0 {
        date += Duration::days(step as i64);
        if is_trading_date_naive(date) {
            remaining -= 1;
        }
    }
    date
}

pub(crate) fn is_market_holiday_naive(date: NaiveDate) -> bool {
    holiday_dates_for_year(date.year()).contains(&date)
}

pub(crate) fn is_early_close_naive(date: NaiveDate) -> bool {
    early_close_dates_for_year(date.year()).contains(&date)
}

pub(crate) fn is_trading_date_naive(date: NaiveDate) -> bool {
    is_weekday(date) && !is_market_holiday_naive(date)
}

pub fn trading_day_info(date: &str) -> TimeResult<TradingDayInfo> {
    let date = parse_naive_date(date)?;
    let is_trading_date = is_trading_date_naive(date);
    let is_market_holiday = is_market_holiday_naive(date);
    let is_early_close = is_trading_date && is_early_close_naive(date);
    let market_hours = market_hours_for_date(&date.format("%Y-%m-%d").to_string())?;

    Ok(TradingDayInfo {
        date: date.format("%Y-%m-%d").to_string(),
        is_trading_date,
        is_market_holiday,
        is_early_close,
        market_hours,
    })
}

pub fn is_trading_date(date: &str) -> bool {
    parse_naive_date(date).map(is_trading_date_naive).unwrap_or(false)
}

pub fn is_trading_today() -> bool {
    is_trading_date(&today())
}

pub fn market_hours_for_date(date: &str) -> TimeResult<MarketHours> {
    let date = parse_naive_date(date)?;
    let date_string = date.format("%Y-%m-%d").to_string();
    let is_trading_date = is_trading_date_naive(date);
    let is_early_close = is_trading_date && is_early_close_naive(date);

    if !is_trading_date {
        return Ok(MarketHours {
            date: date_string,
            is_trading_date: false,
            is_early_close: false,
            premarket_open: None,
            regular_open: None,
            regular_close: None,
            after_hours_close: None,
        });
    }

    Ok(MarketHours {
        date: date_string,
        is_trading_date: true,
        is_early_close,
        premarket_open: Some("04:00".to_string()),
        regular_open: Some("09:30".to_string()),
        regular_close: Some(if is_early_close { "13:00" } else { "16:00" }.to_string()),
        after_hours_close: Some("20:00".to_string()),
    })
}

pub fn last_completed_trading_date(at_timestamp: Option<&str>) -> TimeResult<String> {
    let timestamp = parse_naive_timestamp(at_timestamp.unwrap_or(&now()))?;
    let date_string = timestamp.format("%Y-%m-%d").to_string();
    let hours = market_hours_for_date(&date_string)?;

    if !hours.is_trading_date {
        return add_trading_days(&date_string, -1);
    }

    let regular_close = hhmm_to_total_minutes(
        hours
            .regular_close
            .as_deref()
            .ok_or_else(|| TimeError::new("invalid_market_hours", "missing regular close"))?,
    )?;
    let current_minutes = timestamp.hour() * 60 + timestamp.minute();

    if current_minutes >= regular_close {
        Ok(date_string)
    } else {
        add_trading_days(&date_string, -1)
    }
}

pub fn add_trading_days(date: &str, days: i32) -> TimeResult<String> {
    let date = parse_naive_date(date)?;
    Ok(add_trading_days_from(date, days)
        .format("%Y-%m-%d")
        .to_string())
}

pub(crate) fn hhmm_to_total_minutes(hhmm: &str) -> TimeResult<u32> {
    minutes_from_hhmm(hhmm)
}
